### PR TITLE
Suggest adding a semicolon when unit type is expected

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

Relates to #395 in that it suggests a solution, though this does not explain why for loops usually need semicolons in noir.

# Description

## Summary of changes

Adds a suggestion when a user gets a `expected Unit` type error within blocks to add a semicolon at the end of a statement. This is most often encountered for for loops currently.

Edit: I have also added a more specific hint for function calls. Here are what the two new hints look like:
```rs
error: Expected type (), found type u32
   ┌─ /Users/jakefecher/Code/Noir/noir/crates/nargo/tests/test_data/8_integration/src/main.nr:11:5
   │
11 │     siggy(3);
   │     --------
   │
   = Function calls in noir have no side effects. Perhaps you meant to bind the result to a name?

error: Expected type (), found type [()]
   ┌─ /Users/jakefecher/Code/Noir/noir/crates/nargo/tests/test_data/8_integration/src/main.nr:17:5
   │  
17 │ ╭     for i in 0..10 {
18 │ │         for j in 0..10 {
19 │ │             c[i+10*j] = 0 as u32;
20 │ │ 
21 │ │             for k in 0..10 {
22 │ │                c[i+10*j] = c[i+10*j] + a[i+10*k]*b[k+10*j];
23 │ │             };
24 │ │         };
25 │ │     }
   │ ╰─────'
   │  
   = Try adding a semicolon after this statement
```

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

### CODE REVIEW

**Review by vezenovm — APPROVED** [review]
Looks good

File: crates/noirc_frontend/src/hir/type_check/expr.rs [hunk 1]
```diff
@@ -274,6 +270,39 @@ fn type_check_index_expression(
...
+        expected_typ: Type::Unit.to_string(),
+        expr_typ: expr_type.to_string(),
+        expr_span: span,
+    };
+    match interner.statement(stmt) {
+        HirStatement::Semi(_) => err,
+        HirStatement::Expression(id) => {
+            match interner.expression(&id) {
+                HirExpression::Call(_) | HirExpression::MethodCall(_) => {
+                    err.add_context("Function calls in noir have no side effects. Perhaps you meant to bind the result to a name?")
```
Comment (guipublic) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:297]: Function calls in noir do have side effect, we should allow users to call a function.

Comment (vezenovm) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:297]: I was under the impression that this error only happens if the function a user is calling has a return type but they have not assigned it to a variable. For example, in the tests where functions return `()`, the functions are still able to be called (https://github.com/noir-lang/noir/blob/21414180bd2f5a2d5f63bc75ee61a8d5e1a13aa6/crates/nargo/tests/test_data/global_consts/src/main.nr#L41).

Comment (guipublic) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:297]: Yes, this happens with functions having a return value. But it should not be an error.

**Comment by jfecher** [comment]
I think this should be a warning by default as well. Though warnings as a whole are treated as errors by default currently.

**Comment by kevaundray** [comment]
In an internal meeting, I think we came to the general consensus that we were okay with changing for expressions into for statements

**Comment by vezenovm** [comment]
We could probably close this PR as it is handled by PR #514 ?

**Comment by jfecher** [comment]
It is not handled by PR #514, this PR is on adding a hint to add a semicolon in error messages when a unit type is expected.

**Comment by vezenovm** [comment]
> It is not handled by PR #514, this PR is on adding a hint to add a semicolon in error messages when a unit type is expected.

Ah yes, I was thinking as for-exprs no longer require a semicolon, but I guess this would apply to more use cases than just for-exprs.

**Comment by kevaundray** [comment]
@jfecher what is the status with this PR?

**Comment by jfecher** [comment]
We can close it. It has rotted somewhat with for loops now returning unit by default anyway. The semicolon suggestion is a one line change we can reintroduce if needed.

**Pull request closed without merging**